eli5/keras.py

```diff
@@ -0,0 +1,293 @@
+
+import numpy as np
+import keras
+import keras.backend as K
+from keras.models import Model
+from keras.layers import Layer, Lambda
+from keras.preprocessing.image import load_img, img_to_array, array_to_img
+
+from eli5.base import Explanation
+from eli5.explain import explain_prediction
+
+
+DESCRIPTION_KERAS = """Grad-CAM visualization for image input; output is images"""
+
+# note that keras.models.Sequential subclasses keras.models.Model, so we can just register Model
+@explain_prediction.register(Model)
+def explain_prediction_keras(estimator, doc,
+                             target_names=None, # TODO: implement this
+                             targets=None,
+                             # new parameters:
+                             layer=None,
+                            ):
+    """
+    Explain an image prediction of a Keras image classifier.
+
+    See :func:`eli5.explain_prediction` for more information about the ``estimator``,
+    ``doc``, ``target_names``, and ``targets`` parameters.
+
+    Parameters
+    ----------
+    estimator : object
+        Instance of a Keras neural network model.
+    doc : object
+        An input image as a tensor to ``estimator``, for example ``numpy.ndarray``.
+    target_names : list, optional
+        *`Not Implemented`*.
+
+        Names for classes in the final output layer.
+    targets : list[int], optional
+        Prediction ID's to focus on.
+
+        *`Currently only the first prediction from the list is explained`*.
+
+        If None, the model is fed the input and its top prediction 
+        is taken as target as the target automatically.
+    layer : int or str or object, optional
+        The activation layer in the model to perform Grad-CAM on,
+        a valid keras layer name, layer index, or an instance of keras.layers.Layer.
+        
+        If None, a suitable layer is attempted to be retrieved (raise ValueError if can not).
+
+    Returns
+    -------
+    A :class:`base.Explanation` object with the ``image`` and ``heatmap`` attributes set.
+    """
+    activation_layer = get_activation_layer(estimator, layer)
+    predicted = get_target_prediction(estimator, doc, targets)
+    
+    heatmap = grad_cam(estimator, doc, predicted, activation_layer)
+    # TODO: consider renaming 'heatmap' to 'visualization'/'activations' (the output is not yet a heat map)
+
+    # need to insert a 'channel' axis to get a rank 3 image from rank 2
+    heatmap = np.expand_dims(heatmap, axis=-1)
+    heatmap = array_to_img(heatmap) # -> PIL image
+    
+    # take the single image from the input 'batch' (rank 4 to rank 3)
+    doc = doc[0]
+    image = array_to_img(doc) # -> PIL image
+
+    # TODO: return arrays, not images (arrays are more general)
+    return Explanation(
+        estimator.name, # might want to replace this with something else, eg: estimator.summary()
+        description=DESCRIPTION_KERAS,
+        error='',
+        method='Vanilla Grad-CAM',
```

Comment:
I think it's fine to call it just "Grad-CAM"

Reply:
https://github.com/TeamHG-Memex/eli5/pull/315/commits/9744451a5d3ddb137c5631fde450f2197656d442